### ISSUE TITLE
[top / edn] inter-module broadcast bus 

### ISSUE THREAD

**Issue by tjaychen** [lowRISC/opentitan#4447]
The inter-module function supports the broadcast function where a single port on a particular instance is connected to ports of multiple instances (see `csrng`).  However, it doesn't seem to support the case if only a partial number of the connections are populated.

As an example, `edn` is able to support 4 connections. However, when hooking up only one, the other 3 seem to just be left dangling.  This might be a problem because the total number of connection points on `edn` is fixed, so if we don't have EXACTLY the same number of points (and we might not) something might be left dangling. 

I think we can do some pretty simple fixes such as declaring default values for those connections and just overriding the rest in an always block.  Of course there may be other fixes as well.

**Comment by tjaychen** [comment]
FYI
@msfschaffner 
@senelson7

**Comment by mwbranstad** [comment]
@tjaychen I am not sure I would call the csrng app i/f a "broadcast" function. It is multiple but unique instances of a specific interface. For EDN, I was thinking that we just set the NumEndpoints parameter once the topology of the end points has been determined, in which case there will be no "danglers". Does this make sense? BTW, defaults are being set (last I checked) to unused bus ports. In such a case I would expect synthesis to remove this logic.

**Comment by tjaychen** [comment]
sorry, the "broadcast" term here is somewhat specific to inter-module. It just means a port on one instance connects to an array of instances, like the way csrng is connected to both `edn`'s.

So what i'm saying in this case .. let's say csrng is hardwired to connect to 3 end points (making this up).  However, in `top_earlgrey.hjson` we only specify two connection points, either because the 3rd is not yet there or doesn't exist.  The generated output just leaves that last one dangling and doesn't hook up a default for that one while wiring up the other two.

Does that make sense?

**Comment by mwbranstad** [comment]
Yes, i thought that was what you were saying. So far I have just hooked up two EDN blocks to CSRNG with no danglers (at least that was the intent). If we want to add more, we need to change all of the parameters, etc. to match the new topology. Doesn't that seem reasonable?

**Comment by vogelpi** [comment]
If I understand it correctly, the question is not about what is connected to CSRNG, but what is connected to EDN. Currently, we have:
```
      /- EDN0
CSRNG-|
      \- EDN1
```
So nothing is connected to the EDNs. The peripheral ports on the EDNs are all "terminated" meaning all request inputs are tied to 0.

Now, if we connect a single peripheral, say keymgr, to EDN0, we end up with:
```
              /- keymgr
      /- EDN0 -- dangling
      |      |-- dangling 
      |       \- dangling
CSRNG-|
      \- EDN1
```
So only peripheral port 0 is connected. The other ports on EDN0 are dangling, the corresponding request signals are not defined. And this is potentially problematic.

**Comment by mwbranstad** [comment]
hi @vogelpi your ascii diagrams appear accurate. My take away from the second scenario is that if we hook up one endpoint, the default input settings are not applied to the other ports, and X's will propagate in that case. If that is the case, I would recommend we build endpoint stub blocks and use them until all the real blocks are connected. Due to possible interconnect limitations, the stub blocks may have to be used in the long term as well.

**Comment by tjaychen** [comment]
i just talked to @eunchan about this, and he actually suggested a different method we can use to connect up the edn points.
We can go that approach for now and maybe plan for enhancing the inter-module flow later to natively cover dangling points.

Would you want to have a look at that @mwbranstad ? :)

**Comment by tjaychen** [comment]
@eunchan actually it seems like the `inter-module` `top` type assumes that the module signal pairing already exists.
I think what I want is a bit different... ie.. i want to be able to define a "dummy net" for one of the connections, while the others are real.

So i'm thinking about adding an `inter-module net` type... and the only difference between this one and `top` is that it's not associated with a module.. and other connections can search for it.  What do you think?

I originally wanted to have edn directly drive a `top` type, but that would mean we need hard-wire the connections to the other modules right? Or at a minimum make the top type searchable by the `connect` type.

**Comment by eunchan** [comment]
Right, if we want to use `top` type, then the other end modules also defines its IM port in `top` type and the template needs the connections.

**Comment by tjaychen** [comment]
addressed by #5107
